cmd/nexusUpload.go

```diff
@@ -0,0 +1,145 @@
+package cmd
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"hash"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+
+	"crypto/md5"
+	"crypto/sha1"
+
+	piperHttp "github.com/SAP/jenkins-library/pkg/http"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/telemetry"
+)
+
+type artifactDescription struct {
+	ID         string `json:"artifactId"`
+	Classifier string `json:"classifier"`
+	Type       string `json:"type"`
+	File       string `json:"file"`
+}
+
+func nexusUpload(config nexusUploadOptions, telemetryData *telemetry.CustomData) {
+	artifacts := getArtifacts(config)
+	baseUrl := getBaseUrl(config)
+	client := createHttpClient(config)
+
+	for _, artifact := range artifacts {
+		url := getArtifactUrl(baseUrl, config.Version, artifact)
+
+		uploadHash(client, artifact.File, url+".md5", md5.New(), 16)
+		uploadHash(client, artifact.File, url+".sha1", sha1.New(), 20)
+		uploadFile(client, artifact.File, url)
+	}
+}
+
+func getArtifacts(config nexusUploadOptions) []artifactDescription {
+	var artifacts []artifactDescription
+	err := json.Unmarshal([]byte(config.Artifacts), &artifacts)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("Failed to convert artifact JSON '", config.Artifacts, "'")
+	}
+	return artifacts
+}
+
+func createHttpClient(config nexusUploadOptions) *piperHttp.Client {
+	client := piperHttp.Client{}
+	clientOptions := piperHttp.ClientOptions{Username: config.User, Password: config.Password, Logger: log.Entry().WithField("package", "github.com/SAP/jenkins-library/pkg/http")}
+	client.SetOptions(clientOptions)
+	return &client
+}
+
+func getBaseUrl(config nexusUploadOptions) string {
+	baseUrl := config.Url
+	switch config.NexusVersion {
+	case "nexus2":
+		baseUrl += "/content/repositories/"
+	case "nexus3":
+		baseUrl += "/repository/"
+	default:
+		log.Entry().Fatal("Unsupported Nexus version '", config.NexusVersion, "'")
+	}
+	groupPath := strings.ReplaceAll(config.GroupID, ".", "/")
+	baseUrl += config.Repository + "/" + groupPath + "/"
+	return baseUrl
+}
+
+func getArtifactUrl(baseUrl, version string, artifact artifactDescription) string {
+	url := baseUrl
+
+	// Generate artifacte name including optional classifier
+	artifactName := artifact.ID + "-" + version
+	if len(artifact.Classifier) > 0 {
+		artifactName += "-" + artifact.Classifier
+	}
+	artifactName += "." + artifact.Type
+
+	url += artifact.ID + "/" + version + "/" + artifactName
+
+	// Remove any double slashes, as Nexus does not like them, and prepend protocol
+	url = "http://" + strings.ReplaceAll(url, "//", "/")
```

Comment:
I would like to propose not to hardcode the protocol here, since it is likely to have also `https` use cases here. Plain http communication is more and more vanishing. I remember a [case](https://github.com/SAP/devops-docker-images/pull/130) we had recently where we had to apply an update since e.g. maven central does not support the good old plain `http` protocol anymore. One option is: let the baseUrl already contain the protocol. Other option: provide a parameter in the configuration

Reply:
OK, will go with the second suggestion, thanks!